test/unified_format.py

```diff
@@ -0,0 +1,999 @@
+        super(UnifiedSpecTestMeta, cls).__init__(*args, **kwargs)
+
+        def create_test(spec):
+            def test_case(self):
+                self.run_scenario(spec)
+            return test_case
+
+        for test_spec in cls.TEST_SPEC['tests']:
+            description = test_spec['description']
+            test_name = 'test_%s' % (description.strip('. ').
+                                     replace(' ', '_').replace('.', '_'),)
+            test_method = create_test(copy.deepcopy(test_spec))
+            test_method.__name__ = test_name
+
+            if description in cls.EXPECTED_FAILURES:
+                test_method = unittest.expectedFailure(test_method)
+
+            setattr(cls, test_name, test_method)
+
+
+_ALL_MIXIN_CLASSES = [
+    UnifiedSpecTestMixinV1,
+    # add mixin classes for new schema major versions here
+]
+
+
+_SCHEMA_VERSION_MAJOR_TO_MIXIN_CLASS = {
+    KLASS.SCHEMA_VERSION[0]: KLASS for KLASS in _ALL_MIXIN_CLASSES}
+
+
+def generate_test_classes(test_path, module=__name__, class_name_prefix='',
+                          expected_failures=[]):
+    """Method for generating test classes. Returns a dictionary where keys are
+    the names of test classes and values are the test class objects."""
+    test_klasses = {}
+
+    def test_base_class_factory(test_spec):
+        """Utility that creates the base class to use for test generation.
+        This is needed to ensure that cls.TEST_SPEC is appropriately set when
+        the metaclass __init__ is invoked."""
+        class SpecTestBase(with_metaclass(UnifiedSpecTestMeta)):
+            TEST_SPEC = test_spec
+            EXPECTED_FAILURES = expected_failures
+        return SpecTestBase
+
+    for dirpath, _, filenames in os.walk(test_path):
+        dirname = os.path.split(dirpath)[-1]
+
+        for filename in filenames:
+            fpath = os.path.join(dirpath, filename)
+            with open(fpath) as scenario_stream:
+                # Use tz_aware=False to match how CodecOptions decodes
+                # dates.
+                opts = json_util.JSONOptions(tz_aware=False)
+                scenario_def = json_util.loads(
+                    scenario_stream.read(), json_options=opts)
+
+            test_type = os.path.splitext(filename)[0]
+            snake_class_name = 'Test%s_%s_%s' % (
+                class_name_prefix, dirname.replace('-', '_'),
+                test_type.replace('-', '_').replace('.', '_'))
+            class_name = snake_to_camel(snake_class_name)
+
+            try:
+                schema_version = Version.from_string(
+                    scenario_def['schemaVersion'])
+                mixin_class = _SCHEMA_VERSION_MAJOR_TO_MIXIN_CLASS.get(
+                    schema_version[0])
+                if mixin_class is None:
+                    print('Ignoring test file %s with '
+                          'unsupported schemaVersion %s' %
+                          (fpath, schema_version))
```

Comment:
Shouldn't we raise some kind of error here so that we actually notice when this happens?

Reply:
If we do that, we cannot get the test suite green with the `valid-fail` tests: https://github.com/mongodb/specifications/tree/master/source/unified-test-format/tests

Reply:
Shouldn't the `valid-fail` tests catch said error to assert that the tests actually do fail?

Reply:
Like if we copy/paste this test file https://github.com/mongodb/specifications/blob/master/source/unified-test-format/tests/valid-fail/schemaVersion-unsupported.yml into the `valid-pass` directory, and then run the test suite, it should fail. What's the behavior in that scenario right now?

Reply:
I would think if we expect something to fail, we would mark it as an expected failure and have the test suite still pass. This is what we are doing at the moment with the `valid-fail` tests (see the `expected_failures` argument): 
```
globals().update(generate_test_classes(
    os.path.join(_TEST_PATH, 'valid-fail'),
    module=__name__,
    class_name_prefix='UnifiedTestFormat',
    expected_failures=[
        'foo',
        'FindOneAndReplace returnDocument invalid enum value',
        'FindOneAndUpdate returnDocument invalid enum value'
    ]))
```

IIRC the spec is not prescriptive for this. We need to decide what the correct behavior should be.

Reply:
I don't think we should need to list out the `expected_failures`. I think it's clear that if a test is in the `valid-fail` directory it should fail and any test in `valid-pass` should pass. Is that the question? Or is the question what type of error is expected?

I'd like to avoid squashing errors and instead raise them.

Reply:
Updated the code to raise errors. The unsupported schema test file now causes a failure during test class generation as follows: 
```
(pymongo_cp37) ➜  mongo-python-driver git:(PYTHON-2033/unified-test-format) ✗ python setup.py test -s test.test_unified_format
running test
running egg_info
writing pymongo.egg-info/PKG-INFO
writing dependency_links to pymongo.egg-info/dependency_links.txt
writing requirements to pymongo.egg-info/requires.txt
writing top-level names to pymongo.egg-info/top_level.txt
reading manifest file 'pymongo.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
writing manifest file 'pymongo.egg-info/SOURCES.txt'
running build_ext
Traceback (most recent call last):
  File "setup.py", line 431, in <module>
    **extra_opts
  File "/Users/pmital/.pyenv/versions/3.7.5/envs/pymongo_cp37/lib/python3.7/site-packages/setuptools/__init__.py", line 145, in setup
    return distutils.core.setup(**attrs)
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "setup.py", line 123, in run
    self.test_suite)
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "/Users/pmital/Developer/mongo-python-driver/test/test_unified_format.py", line 46, in <module>
    'FindOneAndUpdate returnDocument invalid enum value'
  File "/Users/pmital/Developer/mongo-python-driver/test/unified_format.py", line 989, in generate_test_classes
    fpath, schema_version))
ValueError: test file '/Users/pmital/Developer/mongo-python-driver/test/unified-test-format/valid-fail/schemaVersion-unsupported.json' has unsupported schemaVersion '0.1.0.0'
```

Note that with this change I have had to comment out the code that generates the `valid-fail` tests since leaving them would break CI (as they fail when run).

Reply:
The other issue to consider is that there is now no way to run the other tests in the `valid-fail` directory without removing the `schemaVersion-unsupported.json` file. Previously, we continued to create test classes even after encountering a failure. Since we are now eagerly surfacing these errors to the user, this is no longer possible. 
I don't think it will be a problem in practice (actual spec tests will never cause such a failure, and I think we'd be OK seeing a build failure if they did), but I thought I should point it out nonetheless.